openfaas-incubator/openfaas-operator
# Invoke via endpoints lister

## Description


Invoke via endpoints lister

## Motivation and Context



Fixes: #102 

This patch makes the operator use endpoint load-balancing using
a random distribution and is a copy of the approach used in
faas-netes. It is used on an opt-in basis depending on whether the
gateway is set to `direct_functions=false`. 

## How Has This Been Tested?
    
At present, it fails with "error listing" despite updating local

Simply:

* Deploy the new version of the operator:

```
helm repo update && helm upgrade openfaas --install openfaas/openfaas    \
 --namespace openfaas    \
  --set basic_auth=true  \
   --set functionNamespace=openfaas-fn --set clusterRole=false \
  --set operator.create=true --set gateway.directFunctions=false \
  --set gateway.image=openfaas/gateway:0.18.3 \ 
  --set operator.image=openfaas/openfaas-operator:0.12.1-rc5
```

* Set `openfaas/openfaas-operator:0.12.1-rc5` with your own image

* Invoke the function

* See the error
## Types of changes


## Impact to existing users


Disable `direct_functions` in the gateway.

Sample image for testing: `openfaas/openfaas-operator:0.12.1-rc1`

RBAC permissions also need to be edited to include endpoints.

Edit the role: `kubectl edit -n openfaas-fn role/openfaas-operator-rw`

Add as per:

```yaml
- apiGroups:
  - ""
  resources:
  - pods
  - pods/log
  - namespaces
  - endpoints
  verbs:
  - get
  - list
  - watch
```
